### ISSUE TITLE
amazon bot failed to execute after item was found

### ISSUE THREAD

**Issue by drx19** [Hari-Nagarajan/nvidia-bot#201]
![image](https://user-images.githubusercontent.com/71615615/94627102-a1e7d680-028a-11eb-9bc5-1a74f00d8f54.png)

**Comment by drx19** [comment]
I have a theory but not confirmed to why the bot failed. I have been getting page not found on many amazon pages after running the bot, including my account page and even searches now I get the following page not found error 
![image](https://user-images.githubusercontent.com/71615615/94627494-b6789e80-028b-11eb-8192-8a845c7138f9.png)

**Comment by drx19** [comment]
update: clearing the browser info cache and cookies brought amazon back. not sure if the bot will function as expected or if this was the reason it failed

**Comment by jcee14** [comment]
I just had the same error for amazon us (https://www.amazon.com/gp/product/B08HJTH61J/) 
I managed to buy manually, so item was definitely in stock.  Unfortunately I don't have a screengrab of the browser window at the time, but I recall seeing a normal Amazon page no with no error.  Just seemed like the buttons were not located where the bot expected them to be.
```
selenium.common.exceptions.NoSuchElementException: Message: no such element: Unable to locate element: {"method":"xpath","selector":"//*[@id="sc-buy-box-ptc-button"]/span/input"}
```

**Comment by drod3763** [comment]
I talked about this in the Discord.

There are some scenarios where the amazon bot will find availability for an item and then the item is removed right by amazon right after clicking the continue button. It's either a "False" item in stock or maybe availability that shows up a few seconds before it's actually available. If this happens when it click Continue and it doesn't find the checkout button it crashes.

Instead of crashing, we should go back to trying to add to cart again. I dont know python, but maybe surrounding this with a try-catch, and going back one step when the NoSuchElementException is thrown or simply just handling the null case.

**Comment by jcee14** [comment]
I have been running the bot with 6fec3df30fd7f6ff7ba715740878d29170f9066e reverted which prevents diagnosing the problem, and I found one problematic scenario. 
![unable_to_buy](https://user-images.githubusercontent.com/6411280/94983714-70ce0700-0513-11eb-8520-e06831845dc9.png)

Since it is asking "Please confirm that you want to add the item(s) listed below to your cart.", I am assuming you need to click the Continue button before you are able to continue with the checkout process.

HTML for the button that would need to be clicked:
```
<input type="image" src="https://images-na.ssl-images-amazon.com/images/G/01/x-locale/communities/reminder/v2/continue._CB485934495_.gif" width="94" alt="Continue" value="add" name="add" height="22" border="0">
```

**Comment by drod3763** [comment]
That's how it works already the problem is what happens after the Continue button is clicked, it sometimes results in an empty cart. That means there is no "Checkout" button which causes it to crash.

**Comment by jcee14** [comment]
There certainly may be multiple issues... not sure if the most recent time I encountered this error is the same as the first time.  The thing about the first time I ran into this error is that I was able to **manually check out the item after the bot crashed**.  It's super backordered, not sure if that is evidence of it being a separate batch than the initial stock or whether checkout failed for some reason.

**Comment by jcee14** [comment]
> HTML for the button that would need to be clicked:
> 
> ```
> <input type="image" src="https://images-na.ssl-images-amazon.com/images/G/01/x-locale/communities/reminder/v2/continue._CB485934495_.gif" width="94" alt="Continue" value="add" name="add" height="22" border="0">
> ```
Looking at the code, I do see that there is supposed to click it.  Not sure from reading the code whether it waits for the page to change to the cart page though.

**Comment by drod3763** [comment]
It should have a built-in wait time for that if im not mistaken. The bot definitely works as I've been using it to buy some hard to obtain cleaning supplies. The issue is perhaps that it's clicking Continue too fast. 

To rephrase what I said before is that what I've been seeing is that there is sometimes a short period between the item being available to be added to the cart but it is actually not yet in stock to be purchased. If you click Continue too fast, it never gets added to the Cart, there is no Checkout button and... exception.

This is something humans don't experience because we don't often move that fast through the site. The best solution I could think of was to go back to the cart and try again. That way it should still work eventually if the stock actually gets recorded. 

Assuming the error you posted is the same one that led to the screenshot you saw - it might be that this scenario is a situation where it ran into the error, went back to the stock check and tried to click checkout on the wrong page. Unless the xpath is the same for the "Continue" and "Checkout" buttons, then I dunno.

**Comment by jcee14** [comment]
Perhaps your theory is correct, hard to validate right now *shrug*.

I excluded the code which has the bot go back to the stock check by reverting https://github.com/Hari-Nagarajan/nvidia-bot/commit/6fec3df30fd7f6ff7ba715740878d29170f9066e so I'm fairly certain the screenshot I posted is the state of the page when the bot crashed (excluding any transient states at page load time).

**Comment by drod3763** [comment]
For science I will do the same. It would have been nice to see the bot logs and the stack trace you had though.

**Comment by jcee14** [comment]
It is identical to the original post

**Comment by jcee14** [comment]
For science! ✨

**Comment by drod3763** [comment]
If its identical then it definitely clicked Continue. Maybe the Continue button wasn't loaded when it tried, or the click failed register for some reason (maybe clicked before the page finished loading). The bot thought it was on the Checkout page but it wasn't, tried to click it and then failed. 

Maybe this can be solved by verifying the page that you are on is the correct one. If you are on the add to cart page, then you click Continue. If you are still on the cart page after that then it should throw an error.

Once we confirmed that is happening, we can fix it by just clicking Continue again after a second or something like that.

**Comment by jcee14** [comment]
Yeah I feel like those are likely scenarios.  Looking at the following code

What's interesting to note is that 
```
log.info("Waiting for Cart Page")
self.check_if_captcha(self.wait_for_pages, SHOPING_CART_TITLES)
```
needs to happen before it fails to click the checkout button. 
However, the page in my screenshot does not match any of the titles in `SHOPING_CART_TITLES`

**Comment by camborambo** [comment]
I just had something similar happen about an hour ago. It would add to cart and hit continue but the error happens because the shopping cart is empty after the click.

**Comment by drod3763** [comment]
That should be handled by the bugfix i think, @jcee14 is seeing something I might have seen before too. That should probably be a new issue though.

**Comment by camborambo** [comment]
> 
> 
> That should be handled by the bugfix i think, @jcee14 is seeing something related but slightly different. I think I might have seen it before too. That should probably be a new issue though.

Yea, I am trying to upload the image I have but it's not working for some reason. I'll give it a try tomorrow or upload. It is getting too late haha.
